File: ruby/raindrops/raindrops.rb
class Raindrops
  FACTOR_STRINGS = {
    3 => 'Pling',
    5 => 'Plang',
    7 => 'Plong'
  }

  def self.convert n
    s = stringify(factors(n))
    s.empty? ? n.to_s : s
  end

  private
    def self.stringify(factors)
      factors.map { |factor| FACTOR_STRINGS[factor] }.join
    end

    def self.factors(n)
      (1..n).select { |i| (n % i).zero? }
    end
end

module BookKeeping
  VERSION = 3
end

File: go/raindrops/raindrops.go
// Package raindrops implements rain sounds conversion
package raindrops

import (
	"strconv"
)

type rainDrop struct {
	Factor int
	Sound  string
}

var drops = []rainDrop{
	{Factor: 3, Sound: "Pling"},
	{Factor: 5, Sound: "Plang"},
	{Factor: 7, Sound: "Plong"},
}

// Convert returns matched rain sounds to given number's factors
func Convert(i int) (s string) {
	for _, drop := range drops {
		if i%drop.Factor == 0 {
			s += drop.Sound
		}
	}
	if s == "" {
		return strconv.Itoa(i)
	}
	return s
}
